packages/eslint-plugin/docs/rules/strict-void-return.mdx

```diff
@@ -0,0 +1,496 @@
+---
+description: 'Disallow passing a value-returning function in a position accepting a void function.'
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/strict-void-return** for documentation.
+
+## Rule Details
+
+TypeScript considers functions returning a value to be assignable to a function returning void.
+Using this feature of TypeScript can lead to bugs or confusing code.
+
+## Examples
+
+### Return type unsafety
+
+Passing a value-returning function in a place expecting a void function can be unsound.
```

Comment:
What do you think of a little more verbosity here? 

If I were writing it, I might do something like this...

````suggestion
Passing a value-returning function in a place expecting a void function can be unsound. TypeScript generally treats the `void` type as though it has the same runtime behavior as `undefined`, so this pattern will cause a mismatch between the runtime behavior and the types.

```ts
// TypeScript errors on overtly wrong ways of populating the `void` type...
function returnsVoid(): void {
    return 'not void!'; // TS Error:  Type 'string' is not assignable to type 'void'.
}

// ...but allows more subtle ones
const returnsVoid: () => void = () => 1234;

// Likewise, TypeScript errors on overtly wrong usages of `void` as a runtime value...
declare const v: void;
if (v) { // TS Error: An expression of type 'void' cannot be tested for truthiness.
    // ... do something
}

// ...but silently treats `void` as `undefined` in more subtle scenarios
declare const voidOrString: void | string;
if (voidOrString) {
     // voidOrString is narrowed to string in this branch, so this is allowed.
     console.log(voidOrString.toUpperCase());
}
```

Putting these behaviors together, we can construct a scenario where TS will throw at runtime, despite not reporting a type error:

````